Added docs for workflow versioning (#5003)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/workflow/workflow-features-concepts.md

## Purging

Workflow state can be purged from a state store, purging all its history and removing all metadata related to a specific workflow instance. The purge capability is used for workflows that have run to a `COMPLETED`, `FAILED`, or `TERMINATED` state.

Learn more in [the workflow API reference guide]({{% ref workflow_api.md %}}).

## Versioning

Workflows can run for very long periods of time, and during that time we want to still be able to introduce changes to the workflow code. When the changes are non-deterministic, we need to use a versioning scheme to version the workflow code so existing workflows can continue running in the old version of the code while new workflows run in the new version of the code.

> Workflows can be versioned in two different ways, each of them with their own advantages and disadvantages. These are:

Reviewer: ```suggestion
Workflows can be versioned in two different ways, each of them with their own advantages. These are:
```

A versioned workflow gets stalled when it started in a new replica but the next replay runs in an old replica.

Stalled workflows will eventually run in a newer replica and will continue.

If a workflow never gets unstalled it means the conditions for getting stalled are still present and will need to be addressed. See the specific reasons for getting stalled in the each of the versioning schemes.

{{% /alert %}}

### Full workflow versioning

This is the traditional way of versioning workflows. The whole workflow code is versioned. This is the easiest way to version workflows and is the recommended way to version workflows.

The different SDKs will expose a way to register multiple handlers for the same workflow name, and will have a way to specify which one is the latest version.

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
registry := task.NewTaskRegistry()
registry.AddVersionedOrchestrator("ProcessOrders", false, OrderProcessingWorkflow)
registry.AddVersionedOrchestrator("ProcessOrders", true, OrderProcessingWorkflowV2)

registry.AddVersionedOrchestratorN("ProcessPayments", "v1", false, ProcessPaymentsWorkflow)

> registry.AddVersionedOrchestratorN("ProcessPayments", "v2", true, ProcessPaymentsWorkflowV2)

Reviewer: Add code comment that `true` and `false` here are signalling latest.

```

**Note**: The boolean argument to `AddVersionedOrchestrator` indicates whether the workflow the latest version.

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, a workflow that started in a specific version will always use the same version of the workflow code even if the workflow code is updated and that version is no longer the latest version.

New workflow instances will always use the latest version of the workflow code.

When the SDK receives a request to run a workflow in a version that is not registered, the workflow will get stalled. As mentioned previously, this can happen naturally during rollouts, but it can also happen if a version is removed while some workflow instances are still using it.

### Patching

> This is a narrower way of versioning workflows, as it allows introducing non-deterministic changes to the workflow code without duplicating the whole workflow code.

Reviewer: Why is the change non deterministic?

Author: Rephrased. I meant that the changes would be non-deterministic if we didn't have this versioning scheme.

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
func UserSignupWorkflow(ctx *task.OrchestrationContext) error {
  // ...
  if ctx.IsPatched("use-sms") {
    if err := ctx.CallActivity("SendSMS", ctx.GetInput()).Await(); err != nil {
      return err
    }
  } else {
    if err := ctx.CallActivity("SendEmail", ctx.GetInput()).Await(); err != nil {
      return err
    }
  }
  // ...
}

```

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, new workflow instances will take the patched code path, but existing workflow instances that are already running at the time the patch is introduced will continue to use the original code path.

The list of patches that are applied to a workflow are stored in the workflow's history, so it's important to be mindful about the amount of patches that are applied to a workflow so the workflow state doesn't grow too large. At this point, consider using Full workflow versioning instead.

There are multiple reasons for workflows to get stalled when using this versioning scheme:
- Removing (or renaming) a patch check.

> - Changing the order of patch checks. It's important to keep the same order of checks throughout the workflow code.

Reviewer: ```suggestion
- Changing the order of patch checks. It's required to keep the same order of checks throughout the workflow code.
```

## Purging

Workflow state can be purged from a state store, purging all its history and removing all metadata related to a specific workflow instance. The purge capability is used for workflows that have run to a `COMPLETED`, `FAILED`, or `TERMINATED` state.

Learn more in [the workflow API reference guide]({{% ref workflow_api.md %}}).

## Versioning

> Workflows can run for very long periods of time, and during that time we want to still be able to introduce changes to the workflow code. When the changes are non-deterministic, we need to use a versioning scheme to version the workflow code so existing workflows can continue running in the old version of the code while new workflows run in the new version of the code.

Reviewer 2: While workflows can certainly run for a long period of time, I'd suggest highlighting this as one of several scenarios instead of articulating it as the only one. Also likely is that workflows are running and mid-flight when an application upgrade is deployed, requiring versioning.

Author: Rephrased, what do you think now?

Workflows can be versioned in two different ways, each of them with their own advantages. These are:
- [Full workflow versioning](#full-workflow-versioning)
- [Patching](#patching)

{{% alert title="Note" color="primary" %}}

> In either of the versioning schemes, workflows can get to the `stalled` state. This usually happens during rollouts, when old and new versions of the workflow code coexist.

Reviewer 2: "can get to" -> "can reach"

> A versioned workflow gets stalled when it started in a new replica but the next replay runs in an old replica.

Reviewer 2: "gets stalled" -> "stalls"
"when it started" -> "when it's started" (consistent tense throughout the sentence)

Stalled workflows will eventually run in a newer replica and will continue.

> If a workflow never gets unstalled it means the conditions for getting stalled are still present and will need to be addressed. See the specific reasons for getting stalled in the each of the versioning schemes.

Reviewer 2: "never gets unstalled" -> "never continues" (follows up on the last sentence)

{{% /alert %}}

> ### Full workflow versioning

Reviewer 2: I'd suggest calling this "Named Workflow Versioning" because @JoshVanL deserves a nice Christmas present

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
registry := task.NewTaskRegistry()
// This is the previous workflow version, so `isLatest` is false
registry.AddVersionedOrchestrator("ProcessOrders", false, OrderProcessingWorkflow)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestrator("ProcessOrders", true, OrderProcessingWorkflowV2)

// This is the previous workflow version, so `isLatest` is false
registry.AddVersionedOrchestratorN("ProcessPayments", "v1", false, ProcessPaymentsWorkflow)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestratorN("ProcessPayments", "v2", true, ProcessPaymentsWorkflowV2)

```

**Note**: The boolean argument to `AddVersionedOrchestratorN` indicates whether the workflow is the latest version.

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, a workflow that started in a specific version will always use the same version of the workflow code even if the workflow code is updated and that version is no longer the latest version.

New workflow instances will always use the latest version of the workflow code.

> When the SDK receives a request to run a workflow in a version that is not registered, the workflow will get stalled. As mentioned previously, this can happen naturally during rollouts, but it can also happen if a version is removed while some workflow instances are still using it.

Reviewer 2: "will get stalled" -> "will stall"

End of the line: "Best practice is that versions shouldn't be deleted at all unless the workflow is no longer in use altogether to avoid such an inadvertent stall."

### Patching

This versioning approach allows you to introduce deterministic changes to a specific part of an existing workflow, rather than creating an entirely new version of the whole workflow. The patching mechanism ensures these targeted changes remain deterministic, whereas without this approach, making such modifications within the workflow code would result in non-determinism. This is especially useful when you want to update or improve just one section of a workflow without duplicating all of its logic.

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
func UserSignupWorkflow(ctx *task.OrchestrationContext) error {
  // ...
  if ctx.IsPatched("use-sms") {
    if err := ctx.CallActivity("SendSMS", ctx.GetInput()).Await(); err != nil {
      return err
    }
  } else {
    if err := ctx.CallActivity("SendEmail", ctx.GetInput()).Await(); err != nil {
      return err
    }
  }
  // ...
}

```

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, new workflow instances will take the patched code path, but existing workflow instances that are already running at the time the patch is introduced will continue to use the original code path.

> The list of patches that are applied to a workflow are stored in the workflow's history, so it's important to be mindful about the amount of patches that are applied to a workflow so the workflow state doesn't grow too large. At this point, consider using Full workflow versioning instead.

Reviewer 2: "At this point, consider" -> "Consider transitioning to a named workflow version and removing all your patches when your patch logic grows unwieldy."

> There are multiple reasons for workflows to get stalled when using this versioning scheme:

Reviewer 2: "to get stalled" -> "to stall"

- Removing (or renaming) a patch check.

> - Changing the order of patch checks. It's required to keep the same order of checks throughout the workflow code.

Reviewer 2: Duplicate patch checks should stall too as it suggests that the check may have moved between replays as well.

Author: It's possible to check the same patch multiple times, as long as the final list of evaluated patches remains consistent across reruns. So, for example, the list can potentially be `"patch1", "patch2", "patch2", "patch3"`

Reviewer 2: Sure, but also the order of the patches, right? This is what I was asking about whether the patches need to be reported as standalone events instead of metadata on the event is that the standalone events would allow us to infer the order in which patches are applied between other events. By just building a list of all the patches observed from the event source, shouldn't we infer that a patch is called out of order (even if just duplicated) if we see a list of "patch1", "patch2", "patch3" and then an ask comes in for "patch1" even though it's already been evaluated and in the event history?

Author: if at one point in the workflow we've seen `patch1,patch2,patch3` and there's a new check for `patch1`, the `is_patched` will resolve to the same outcome as the first time it encountered it, and the SDK will report it to the runtime, so it'll send `patch1,patch2,patch3,patch1`. If in a new run of this workflow instance the reported list is not exactly the same as before, it'll get stalled. It can get new patches though, but the ones that were reported previously need to exist, and in the same order.
Makes sense?

Reviewer 2: Sure, we're on the same page here. We covered this on one of the calls and this is how the .NET SDK approaches it now.

## Purging

Workflow state can be purged from a state store, purging all its history and removing all metadata related to a specific workflow instance. The purge capability is used for workflows that have run to a `COMPLETED`, `FAILED`, or `TERMINATED` state.

Learn more in [the workflow API reference guide]({{% ref workflow_api.md %}}).

## Versioning

There are multiple scenarios where it is necessary to introduce changes to workflow code while workflows are actively running. In cases where these changes are non-deterministic, a versioning scheme is required so that existing workflows can continue executing with the original code, while new workflows use the updated version.

Workflows can be versioned in two different ways, each of them with their own advantages. These are:
- [Full workflow versioning](#full-workflow-versioning)
- [Patching](#patching)

{{% alert title="Note" color="primary" %}}

> In either of the versioning schemes, workflows can reach the `stalled` state. This usually happens during rollouts, when old and new versions of the workflow code coexist.

Reviewer: Please can we have a `dapr workflow list` & `dapr workflow history` command and output section, showing what a stalled workflow looks like.

{{% /alert %}}

### Named Workflow Versioning

This is the traditional way of versioning workflows. The whole workflow code is versioned. This is the easiest way to version workflows and is the recommended way to version workflows.

The different SDKs will expose a way to register multiple handlers for the same workflow name, and will have a way to specify which one is the latest version.

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
registry := task.NewTaskRegistry()
// This is the previous workflow version, so `isLatest` is false
registry.AddVersionedOrchestrator("ProcessOrders", false, OrderProcessingWorkflow)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestrator("ProcessOrders", true, OrderProcessingWorkflowV2)

// This is the previous workflow version, so `isLatest` is false
registry.AddVersionedOrchestratorN("ProcessPayments", "v1", false, ProcessPaymentsWorkflow)
// This is the latest workflow version, so `isLatest` is true

> registry.AddVersionedOrchestratorN("ProcessPayments", "v2", true, ProcessPaymentsWorkflowV2)

Reviewer: Update to use the keyword `Workflow`

New workflow instances will always use the latest version of the workflow code.

When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. As mentioned previously, this can happen naturally during rollouts, but it can also happen if a version is removed while some workflow instances are still using it. Best practice is that versions shouldn't be deleted at all unless the workflow is no longer in use altogether to avoid such an inadvertent stall

### Patching

This versioning approach allows you to introduce deterministic changes to a specific part of an existing workflow, rather than creating an entirely new version of the whole workflow. The patching mechanism ensures these targeted changes remain deterministic, whereas without this approach, making such modifications within the workflow code would result in non-determinism. This is especially useful when you want to update or improve just one section of a workflow without duplicating all of its logic.

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go

> func UserSignupWorkflow(ctx *task.OrchestrationContext) error {

Reviewer: Update to Workflow

registry.AddVersionedOrchestratorN("Workflow", "v1", false, WorkflowV1)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestratorN("Workflow", "v2", true, WorkflowV2)
```

**Note**: The boolean argument to `AddVersionedOrchestratorN` indicates whether the workflow is the latest version.

{{% /tab %}}

{{% tab "Python" %}}

```python
@wfr.versioned_workflow(name="workflow")
def workflow_v1(ctx: wf.DaprWorkflowContext):
  # ...

@wfr.versioned_workflow(name="workflow", is_latest=True)
def workflow_v2(ctx: wf.DaprWorkflowContext):
  # ...

# Or providing a version name

@wfr.versioned_workflow(name="workflow", version_name="v1")
def workflow_v1(ctx: wf.DaprWorkflowContext):
  # ...

@wfr.versioned_workflow(name="workflow", version_name="v2", is_latest=True)
def workflow_v2(ctx: wf.DaprWorkflowContext):
  # ...
```

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, a workflow that started in a specific version will always use the same version of the workflow code even if the workflow code is updated and that version is no longer the latest version.

New workflow instances will always use the latest version of the workflow code.

> When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. As mentioned previously, this can happen naturally during rollouts, but it can also happen if a version is removed while some workflow instances are still using it. Best practice is that versions shouldn't be deleted at all unless the workflow is no longer in use altogether to avoid such an inadvertent stall

Reviewer 3: ```suggestion
When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This can happen naturally during Kubernetes rollouts, but can also happen if a version is removed while some workflow instances are still using it. The best practice is that versions shouldn't be deleted at all unless the workflow is no longer in use altogether to avoid such an inadvertent stall.
```

### Patching

> This versioning approach allows you to introduce deterministic changes to a specific part of an existing workflow, rather than creating an entirely new version of the whole workflow. The patching mechanism ensures these targeted changes remain deterministic, whereas without this approach, making such modifications within the workflow code would result in non-determinism. This is especially useful when you want to update or improve just one section of a workflow without duplicating all of its logic.

Reviewer 3: Can you add a paragraph specifically about checking for specific patches? For instance, reading this I assume, you can check if a specific patch is created then use the new version of the activity based on that return value but its not fully clear. 

Eg. "The following example checks if the `"use-sms"` patch version has been applied to this workflow version. If so then the `"SendSMS"` activity will be executed"

    yield ctx.call_activity(send_sms)
  else:
    yield ctx.call_activity(send_email)
  # ...
```

{{% /tab %}}

{{< /tabpane >}}

In this versioning scheme, new workflow instances will take the patched code path, but existing workflow instances that are already running at the time the patch is introduced will continue to use the original code path.

Patch checks are recorded in the workflow instance history the first time they are evaluated. This means you can safely check the same patch multiple times throughout the workflow, and all checks will resolve the same way for that workflow instance.

The list of patches that are applied to a workflow are stored in the workflow's history, so it's important to be mindful about the amount of patches that are applied to a workflow so the workflow state doesn't grow too large. Consider transitioning to a named workflow version and removing all your patches when your patch logic grows unwieldy.

There are multiple reasons for workflows to stall when using this versioning scheme:
- Removing (or renaming) a patch check.
- Changing the order of patch checks. It's required to keep the same order of checks throughout the workflow code.

> This is what the `dapr workflow list` would look like when a workflow is stalled:

Reviewer 3: Needs more context here. Ie. You can use the Dapr CLI  to check for stalled workflows (link to dapr workflow command). ...

## Purging

Workflow state can be purged from a state store, purging all its history and removing all metadata related to a specific workflow instance. The purge capability is used for workflows that have run to a `COMPLETED`, `FAILED`, or `TERMINATED` state.

Learn more in [the workflow API reference guide]({{% ref workflow_api.md %}}).

## Versioning

> There are multiple scenarios where it is necessary to introduce changes to workflow code while workflows are actively running. In cases where these changes are non-deterministic, a versioning scheme is required so that existing workflows can continue executing with the original code, while new workflows use the updated version.

Reviewer 4: ```suggestion
There are multiple scenarios where it is necessary to introduce changes to workflow code while workflows are actively running. In cases where these changes are non-deterministic, a versioning strategy is required so that existing workflows can continue executing with the original code, while new workflows use the updated version.
```

Workflows can be versioned in two different ways, each of them with their own advantages. These are:
- [Full workflow versioning](#full-workflow-versioning)
- [Patching](#patching)

{{% alert title="Note" color="primary" %}}

> In either of the versioning schemes, workflows can reach the `stalled` state. This usually happens during rollouts, when old and new versions of the workflow code coexist.

Reviewer 4: ```suggestion
In either of the versioning strategies, workflows can reach the `stalled` state. This usually happens during rollouts, when old and new versions of the workflow code coexist.
```

A versioned workflow stalls when it's started in a new replica but the next replay runs in an old replica.

> Stalled workflows will eventually run in a newer replica and will continue.

Reviewer 4: Can you give an indication of time here. It depends on load of course. But if there is very little load what is the order of magnitude for this to resolve? In case a user does not want to wait for automatic resolution of stalled workflow instances how can they solve this issue?

> If a workflow never continues it means the conditions for getting stalled are still present and will need to be addressed. See the specific reasons for getting stalled in the each of the versioning schemes.

Reviewer 4: ```suggestion
If a workflow never continues it means the conditions for getting stalled are still present and will need to be addressed. See the specific reasons for getting stalled in the each of the versioning strategies.
```

{{% /alert %}}

### Named Workflow Versioning

> This is the traditional way of versioning workflows. The whole workflow code is versioned. This is the easiest way to version workflows and is the recommended way to version workflows.

Reviewer 4: ```suggestion
With named workflows the entire workflow code is versioned. This is the easiest way and recommended way to version workflows due to it's simplicity
```

> The different SDKs will expose a way to register multiple handlers for the same workflow name, and will have a way to specify which one is the latest version.

Reviewer 4: ```suggestion
The language SDKs each expose a way to register multiple handlers for the same workflow name, and will have an _is latest_ flag to specify which workflow is the latest version.
```

registry.AddVersionedOrchestrator("Workflow", false, WorkflowV1)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestrator("Workflow", true, WorkflowV2)

// Or providing a version name

// This is the previous workflow version, so `isLatest` is false
registry.AddVersionedOrchestratorN("Workflow", "v1", false, WorkflowV1)
// This is the latest workflow version, so `isLatest` is true
registry.AddVersionedOrchestratorN("Workflow", "v2", true, WorkflowV2)
```

**Note**: The boolean argument to `AddVersionedOrchestratorN` indicates whether the workflow is the latest version.

{{% /tab %}}

{{% tab "Python" %}}

```python
@wfr.versioned_workflow(name="workflow")
def workflow_v1(ctx: wf.DaprWorkflowContext):
  # ...

@wfr.versioned_workflow(name="workflow", is_latest=True)
def workflow_v2(ctx: wf.DaprWorkflowContext):
  # ...

# Or providing a version name

@wfr.versioned_workflow(name="workflow", version_name="v1")
def workflow_v1(ctx: wf.DaprWorkflowContext):
  # ...

@wfr.versioned_workflow(name="workflow", version_name="v2", is_latest=True)
def workflow_v2(ctx: wf.DaprWorkflowContext):
  # ...
```

{{% /tab %}}

{{< /tabpane >}}

> In this versioning scheme, a workflow that started in a specific version will always use the same version of the workflow code even if the workflow code is updated and that version is no longer the latest version.

Reviewer 4: ```suggestion
In this versioning strategy, a workflow that started in a specific version will always use the same version of the workflow code when being replayed even if the workflow code is updated and that version is no longer the latest.
```

New workflow instances will always use the latest version of the workflow code.

> When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This can happen naturally during Kubernetes rollouts, but can also happen if a version is removed while some workflow instances are still using it. The best practice is that versions shouldn't be deleted at all unless the workflow is no longer in use altogether to avoid such an inadvertent stall.

Reviewer 4: ```suggestion
When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This can happen naturally during Kubernetes rollouts, but can also happen if a version is removed while some workflow instances are still using it. The best practice is that versions shouldn't be deleted at all, unless it's guaranteed that there are no more in-flight workflow instances or planned workflow reruns for that version.
```

### Patching

> This versioning approach allows you to introduce deterministic changes to a specific part of an existing workflow, rather than creating an entirely new version of the whole workflow. The patching mechanism ensures these targeted changes remain deterministic, whereas without this approach, making such modifications within the workflow code would result in non-determinism. This is especially useful when you want to update or improve just one section of a workflow without duplicating all of its logic.

Reviewer 4: ```suggestion
The patching strategy allows you to introduce deterministic changes to a specific part of an existing workflow, rather than creating an entirely new version of the entire workflow. The patching mechanism ensures these targeted changes remain deterministic, whereas without this approach, making such modifications within the workflow code would result in non-determinism. This is especially useful when you want to update or improve just one section of a workflow without duplicating all of its logic.
```

{{< tabpane text=true >}}

{{% tab "Go" %}}

```go
func Workflow(ctx *task.OrchestrationContext) error {
  // ...
  if ctx.IsPatched("use-sms") {
    if err := ctx.CallActivity("SendSMS", ctx.GetInput()).Await(); err != nil {
      return err
    }
  } else {
    if err := ctx.CallActivity("SendEmail", ctx.GetInput()).Await(); err != nil {
      return err
    }
  }
  // ...
}

```

{{% /tab %}}

{{% tab "Python" %}}

```python
@wfr.workflow
def workflow(ctx: wf.DaprWorkflowContext):
  # ...
  if ctx.is_patched("use-sms"):
    yield ctx.call_activity(send_sms)
  else:
    yield ctx.call_activity(send_email)
  # ...
```

{{% /tab %}}

{{< /tabpane >}}

> In this versioning scheme, new workflow instances will take the patched code path, but existing workflow instances that are already running at the time the patch is introduced will continue to use the original code path.

Reviewer 4: ```suggestion
In this versioning strategy, new workflow instances will take the patched code path, but existing workflow instances that are in-flight at the time the patch is introduced will continue to use the original code path.
```

Patch checks are recorded in the workflow instance history the first time they are evaluated. This means you can safely check the same patch multiple times throughout the workflow, and all checks will resolve the same way for that workflow instance.

The list of patches that are applied to a workflow are stored in the workflow's history, so it's important to be mindful about the amount of patches that are applied to a workflow so the workflow state doesn't grow too large. Consider transitioning to a named workflow version and removing all your patches when your patch logic grows unwieldy.

> There are multiple reasons for workflows to stall when using this versioning scheme:

Reviewer 4: ```suggestion
There are multiple reasons for workflows to stall when using this versioning strategy:
```

## Purging

Workflow state can be purged from a state store, purging all its history and removing all metadata related to a specific workflow instance. The purge capability is used for workflows that have run to a `COMPLETED`, `FAILED`, or `TERMINATED` state.

Learn more in [the workflow API reference guide]({{% ref workflow_api.md %}}).

## Versioning

> Workflow code is long-running and must remain deterministic during updates. For details on patching and named workflow versioning, see [Workflow versioning]({{% ref workflow-versioning.md %}}).

Reviewer 5: ```suggestion
Workflow code is long-running and must remain deterministic during updates. Workflows support two types of versioning, called patching and named. For details on patching and named workflow versioning, see [Workflow versioning]({{% ref workflow-versioning.md %}}).
```

Reviewer 2: @msfussell  This one feels weird to just say out loud. "Named" feels like it's missing some sort of modifier, which is why I instead opted to describe both as "patch-based versioning" and "named versioning" instead of just "patching" and "named".

Reviewer 5: agreed. Leave this out. Resolved.

deterministically safe way. In this approach, you introduce a new workflow version by duplicating an existing 
workflow and assigning it a new "versioned" name. Because you can then refactor your workflow logic from scratch to 
remove any patches and otherwise introduce any other changes you want, this approach offers a clean break from 
previous workflows versions.

While the particulars of how this is achieved will vary depending on the language SDK you're using, generally, the 
idea is that you'll duplicate your latest workflow, assign it a unique name that reflects the newer version and 
register this with your SDK. A registry will be built of each workflow so that at runtime, Dapr will call in with a 
workflow name to execute and the SDK will either route to the old workflow (if in-flight) or to the newer version.

Do note that across all SDKs, the runtime does not migrate workflows between versions sequentially. Rather, when the 
SDKs receive a request to run a new workflow, it will opt to run the latest version, not simply the "next" one, so 
there's is no need to handle any compensation logic between versions.

> The language SDKs may expsoe a way to register versions when using the same workflow name, but this will vary by SDK,

Reviewer 4: ```suggestion
The language SDKs may expose a way to register versions when using the same workflow name, but this will vary by SDK,
```

so please refer to the documentation for your specific SDK for more information. For example, some may use a mechanism
by which you would explicitly supply an `is_latest` flag to indicate which version is the latest.

When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This 
can happen naturally during application rollouts, but can also happen if a version is removed while some workflow 
instances are still using it. Our recommendation is that older named workflow versions should not be deleted at all 
unless you've independently confirmed that there are no outstanding (in-flight or dormant) workflow instances that 
are running against that version.

{{< tabpane text=true >}}

{{% tab ".NET" %}}

.NET uses a source generator to automatically identify and register your workflow versions, so no manual 

> registration is necessary. By default, .NET uses a built-in and configurably numerical versioning strategy where the 

Reviewer 4: ```suggestion
registration of workflows is necessary (the activities do need to be registered). By default, .NET uses a built-in and configurably numerical versioning strategy where the 
```

> version is provided as a suffix on the name. See the .NET SDK documentation on how to change versioning strategies 

Reviewer 4: Is this already written? If so let's add a link. If not, let's add an issue so we don't forget.

Reviewer 2: Yes, but not merged. As I said before, I intended to come back through here and open a PR pointing to the .NET documentation as necessary once it actually exists, but it's premature right now.

Reviewer 2: Looks like it just [got merged](https://github.com/dapr/docs/pull/5038) yesterday.  @acroca ~~Can you add a link there to `{{% ref dotnet-workflow-versioning %}}`?~~ Could you please grant me permission to push branches on your repo again? I've got a local branch that adds this link, splits versioning out to a separate page (per @msfussell 's ask later in the thread and fixes a typo in the .NET example.

or configuration and how to set up versioning in your application.

Since .NET applications do not allow multiple types to exist with the same name, that's not an option in this SDK.

> To create a new named workflow version, duplicate an existing workflow and modify the name of the class to use the 

Reviewer 4: I do suggest adding a code snippet here that shows `MyWorkflowV1` and `MyWorkflowV2` class definitions to be very explicit.

Reviewer 2: Well, I thought about it, but specifying the workflows alone isn't enough as a DI registration is necessary. I think it makes more sense to point to the SDK documentation specifically, except that .NET hasn't been merged yet, so any such links would be dead. When that PR gets merged, I intended to open another one here to point to those files.

But just specifying workflow types with no other context seems downright silly. At least the other SDKs have annotations they're applying and demonstrating.

Do note that across all SDKs, the runtime does not migrate workflows between versions sequentially. Rather, when the
SDKs receive a request to run a new workflow, it will opt to run the latest version, not simply the "next" one, so
there's is no need to handle any compensation logic between versions.

The language SDKs may expose a way to register versions when using the same workflow name, but this will vary by SDK,
so please refer to the documentation for your specific SDK for more information. For example, some may use a mechanism
by which you would explicitly supply an `is_latest` flag to indicate which version is the latest.

When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This
can happen naturally during application rollouts, but can also happen if a version is removed while some workflow
instances are still using it. Our recommendation is that older named workflow versions should not be deleted at all
unless you've independently confirmed that there are no outstanding (in-flight or dormant) workflow instances that
are running against that version.

{{< tabpane text=true >}}

> {{% tab ".NET" %}}

Reviewer 5: Can we either get this section update based on this https://github.com/dapr/docs/pull/5038 or raise and issue to add the details later

Reviewer 2: @msfussell I modified @acroca 's PR here (via a PR on his own repo) to rewrite this section reflective of both the .NET and other SDKs, so unless it's since changed, it's up to date.

> #### Workflow functions must call deterministic APIs.

Reviewer 2: Remove the period here

{{< /tabpane >}}

> ### Versioning Process Guidance

Reviewer 5: I would either put a link to this section on guidance from the ## Versioning overview, or move this into the  [workflow/workflow-versioning.md] document at the start. It seems odd to keep this on the concepts section otherwise.

> Because named workflows are fully compatible with patches, our intent is that changes to your workflow are initially

Reviewer 5: ```suggestion
Because named workflows are fully compatible with patches, the approach is that changes to your workflow are initially
```

made by applying patches to your existing workflow logic. Eventually you'll hit one of the following problems after
applying several patches:
- You're concerned about the overhead to your workflow state history because of the number of applied patches;
- The golden path through your workflow logic is hard to follow;
- You need to make another workflow tweak, but can't figure out how to apply a patch that doesn't break the
  determinism guarantee.

> At this point, our recommendation is that you introduce a named version of your workflow. Copy your existing

Reviewer 5: ```suggestion
At this point, the recommendation is that you introduce a named version of your workflow. Copy your existing
```

workflow, change its name to reflect whatever versioning strategy you're using in your SDK and refactor to remove all
your patches and retain only your intended "latest" logic. Apply your new changes (no need for a patch here – it's a
new workflow) and deploy it.

Here, apply patches again as needed to address future changes and when necessary, introduce another named workflow
version and continue. Repeat this process indefinitely.

> We recommend that you retain old workflow versions until you're completely confident that nothing is in-flight

Reviewer 5: ```suggestion
It is recommended that you retain old workflow versions until you're completely confident that nothing is in-flight
```

(including deferred with a long-running timer) that uses any of your old workflow versions.

Workflow versioning doesn't address changing the types of inputs and outputs of the workflows themselves. As general

> guidance, we recommend either returning serialized values like strings (making the consumer of the output

Reviewer 5: ```suggestion
guidance, it is recommended to either return serialized values like strings (making the consumer of the output
```
